tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, entry->name().c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+
+    int result;
+    unsigned int firstFrame;
+    unsigned int lastFrame;
+    if ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        firstFrame = sqlite3_column_int(select, 0);
+        lastFrame = sqlite3_column_int(select, 1);
+    }
+
+    ASSERT_SQLITE_DONE(sqlite3_step(select));
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    std::experimental::filesystem::path directory = TileFiles::directoryForTilesInFrames(entry->path(), id, firstFrame, lastFrame);
+
+    return TileFiles::tileFilename(directory, tileNumber);
+}
+
+void LayoutDatabase::addTileLayout(const std::string &video, unsigned int id, unsigned int firstFrame, unsigned int lastFrame, const std::shared_ptr<TileLayout> tileLayout) {
+    addTileBasicLayout(video, id, firstFrame, lastFrame, tileLayout->numberOfColumns(), tileLayout->numberOfRows());
+    addTileLayoutWidths(video, id, tileLayout->widthsOfColumns());
+    addTileLayoutHeights(video, id, tileLayout->heightsOfRows());
+}
+
+void LayoutDatabase::addTileBasicLayout(const std::string &video, unsigned int id, unsigned int firstFrame, unsigned int lastFrame, unsigned int numCols, unsigned int numRows) {
+    std::string query = "INSERT INTO layouts (video, id, first_frame, last_frame, num_rows, num_cols)"\
+                    "VALUES (?, ?, ?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, firstFrame));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, lastFrame));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 5, numRows));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 6, numCols));
+
+    int result = sqlite3_step(select);
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+void LayoutDatabase::addTileLayoutWidths(const std::string &video, unsigned int id, const std::vector<unsigned int> &widths) {
+    std::string query = "INSERT INTO widths (video, id, width, position) "\
+                    "VALUES ( ?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+
+    int position = 0;
+    for (const auto& width : widths) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, width));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
```

Comment:
I think we also want this batch-insert to be inside of a transaction. I can think of a few options:
- We could add a comment saying that this function should only only be called within a transaction (e.g., from `addTileLayout()`)
- We could use sqlite's save points to create nested transactions: https://www.sqlite.org/lang_savepoint.html
- We could insert all of the values at once using an INSERT INTO … VALUES query (there are some examples here: https://www.sqlitetutorial.net/sqlite-insert/)

(Ditto for `addTileLayoutHeights`)